Scanner interface rework

xs5871 opened the pull request:
This replaces the current matrix scanners update object.
Instead of the implementation specific bytearray, there's a `KeyEvent` object inspired by circuitpythons version for their `keypad` library. Main difference to the current implementation is that keys are identified by a single number, instead of row and column.
Outstanding efforts to provide more/different scanners would benefit from this abstraction.

The code has been tested with a monolithic keyboard, and a split using UART. I don't expect there to be issues with BLE, but someone with a wireless split has to confirm.

The only breaking change is in the signature of `intifiy_coordinates`:
```python
intifiy_coordinates(row, column, number_of_columns)
```
I updated the only appearance of ic in the documentation, but I don't own a corne, so I'm half-guessing.
This is also somthing I'd appreciate help verifying.

For reference: this is more or less complimentary to #262. It's been a while since the last update though, and my previous branch is now stale, messy and broke split communication. I understand that this is a big, singular commit, but it's almost impossible to make it more atomic.

klardotsh commented:
if @honboubao (BLE) and @kdb424 (corne) are able to provide manual testing of this, rad, otherwise, this passes my style/flow review and is what I'd consider "fine to merge when tested by at least one other user, or (now that #288 just merged) tests are added to prove it works" (which would be a @xs5871 thing, I imagine)

klardotsh commented:
@kdb424 has merge rights so can green-check and merge this without my further input if they so desire

kdb424 commented:
Not only will the docs have to be updated, but every supported board needs to as well for this to be merged, or anyone using boards like this (lily58, corne, Iris, and more), will be completely broken. Once all of the boards with a non standard matrix are reworked, I will review this.

xs5871 commented:
That's a very good point. Is there a way to change this into a draft for the time being? It'll take me a while to go through all of this.

xs5871 commented:
All boards are ported. There was a mismatch between physical keyboard / keymap layout and `coord_mapping` on the number of keys with two boards. I guessed the correct `coord_mapping` and added a NOQA comment.
I also added a flat `coord_mapping` to the documentation that may be easier to read and understand. The BR translation is computer generated, sorry.

Now to figure out the unit test...

xs5871 commented:
unit test already coming in handy

xs5871 commented:
Verified as working with BLE split with gratitious help from @Tonasz.

kdb424 reviewed (approved):
Thank you both @xs5871 and @Tonasz for your work and testing out code <3
